File: Observer/c++/Observer.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class IObserver
{
    public:
        virtual void update(double p) = 0;
};

class IObservable
{
    public:
        virtual void addObserver(IObserver* o) = 0;
        virtual void removeObserver(IObserver* o) = 0;
        virtual void notify() = 0;
        virtual ~IObservable() {}
};

class Product : public IObservable
{
    private:
        std::vector<IObserver*> observers;
        double price;
        
    public:
        Product(double p) : price(p) {}
        void changePrice(double p);
        void addObserver(IObserver* o) override;
        void removeObserver(IObserver* o) override;
        void notify() override;
};

class Wholesaler : public IObserver
{
    private:
        IObservable* product;

    public:
        Wholesaler(IObservable* obj) : product(obj)
        {
            obj->addObserver(this);
        }
        void update(double p) override;
};

class Buyer : public IObserver
{
    private:
        IObservable* product;

    public:
        Buyer(IObservable* obj) : product(obj)
        {
            obj->addObserver(this);
        }
        void update(double p) override;
};

int main()
{
    Product* product = new Product(400);

    Wholesaler* wholesaler = new Wholesaler(product);
    Buyer* buyer = new Buyer(product);

    product->changePrice(320);
    product->changePrice(280);

    return 0;
}

void Product::changePrice(double p)
{
    price = p;
    notify();
}

void Product::addObserver(IObserver *o)
{
    observers.push_back(o);
}

void Product::removeObserver(IObserver *o)
{
    observers.erase(std::remove(observers.begin(), observers.end(), o), observers.end());
}

void Product::notify()
{
    for (auto o : observers)
    {
        o->update(price);
    }
}

void Wholesaler::update(double p)
{
    if (p < 300)
    {
        std::cout << "the wholesaler bought the goods at a price " << p << std::endl;
        product->removeObserver(this);
    }
}

void Buyer::update(double p)
{
    if (p < 350)
    {
        std::cout << "the buyer purchased the product at the price " << p << std::endl;
        product->removeObserver(this);
    }
}

File: Observer/java/Observer.java
import java.util.ArrayList;
import java.util.List;

interface IObserver{
    void update(double p);
}

interface IObservable{
    void addObserver(IObserver o);
    void removeObserver(IObserver o);
    void updateNotify();
}

class Product implements IObservable{
    private List<IObserver> observers;
    private double price;

    public Product(double p){
        price = p;
        observers = new ArrayList<>();
    }

    public void changePrice(double p){
        price = p;
        updateNotify();
    }

    @Override
    public void addObserver(IObserver o){
        observers.add(o);
    }

    @Override
    public void removeObserver(IObserver o){
        observers.remove(o);
    }

    @Override
    public void updateNotify(){
        for (Object o : observers.toArray()){
            ((IObserver)o).update(price);
        }
    }
}

class Wholesale implements IObserver{
    private IObservable product;

    public Wholesale(IObservable obj){
        product = obj;
        obj.addObserver(this);
    }

    @Override
    public void update(double p){
        if (p < 300){
            System.out.println("the wholesaler bought the goods at a price " + p);
            product.removeObserver(this);
        }
    }
}

class Buyer implements IObserver{
    private IObservable product;

    public Buyer(IObservable obj){
        product = obj;
        obj.addObserver(this);
    }
    
    @Override
    public void update(double p){
        if (p < 350){
            System.out.println("the buyer purchased the product at the price " + p);
            product.removeObserver(this);
        }
    }
}

public class Main {
    public static void main(String[] args){

        Product product = new Product(400);

        Wholesale wholesale = new Wholesale(product);
        Buyer buyer = new Buyer(product);

        product.changePrice(320);
        product.changePrice(280);
    }
}